### PR TITLE
Group name tag for routes and poi

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 1]
```diff
@@ -261,6 +272,32 @@ - (void) populateRouteInfoSection:(OAGPXTableData *)data routeKey:(OARouteKey *)
     [infoSectionData.subjects addObjectsFromArray:subjects];
 }
 
+- (NSArray<NSDictionary *> *)getPoiTypeDataForKey:(NSString *)routeTagKey withValue:(NSString *)value
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:275]: nullable NSDictionary

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 2]
```diff
@@ -261,6 +272,32 @@ - (void) populateRouteInfoSection:(OAGPXTableData *)data routeKey:(OARouteKey *)
...
+{
+    if ([routeTagKey isEqualToString:@"name"])
+        return nil;
+    
+    OAPOIBaseType *poiType = [[OAPOIHelper sharedInstance] getAnyPoiAdditionalTypeByKey:routeTagKey];
+    NSString *localizedTitle = poiType.nameLocalized ?: @"";
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"\\(([^)]+)\\)" options:0 error:nil];
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:282]: add error:&error

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 3]
```diff
@@ -261,6 +272,32 @@ - (void) populateRouteInfoSection:(OAGPXTableData *)data routeKey:(OARouteKey *)
...
+    NSTextCheckingResult *match = [regex firstMatchInString:localizedTitle options:0 range:NSMakeRange(0, localizedTitle.length)];
+    if (match)
+    {
+        NSRange matchRange = [match rangeAtIndex:1];
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:286]: create extension for String. capitalizeFirstLetter

```
 extension NSString {    
    func capitalizingFirstLetter() -> String {
        return prefix(1).capitalized + dropFirst()
    }
    mutating func capitalizeFirstLetter() {
        self = capitalizingFirstLetter()
    }
} 
```

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 4]
```diff
@@ -527,6 +564,22 @@ - (OAGPXTableSectionData *)generateCopyrightSectionData
...
+{
+    NSArray<NSString *> *excludedTags = @[@"int_name", @"nat_name", @"reg_name", @"loc_name", @"old_name", @"alt_name", @"short_name", @"official_name"];
+    NSString *normalizedTagKey = [tagKey stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@":"]];
+    for (NSString *excludedTag in excludedTags)
+    {
+        if ([normalizedTagKey isEqualToString:excludedTag])
+            return YES;
+    }
+    
+    if ([tagKey hasPrefix:@"osmc"] || [tagKey hasPrefix:@"name:"] || ([tagKey isEqualToString:@"relation_id"] && ![OAPluginsHelper isEnabled:OAOsmEditingPlugin.class]))
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:577]: ![OAPluginsHelper isEnabled:OAOsmEditingPlugin.class] - move on start method if needed

File: Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+//  NameTagsDetailsViewController.swift
+//  OsmAnd Maps
+//
+//  Created by Dmitry Svetlichny on 20.06.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import UIKit
+
+@objc(OANameTagsDetailsViewController)
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift:11]: remove

File: Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                guard let tagKey = tag["key"] as? String else { return false }
+                return tagKey.hasPrefix(key)
+            }
+            
+            if !filteredTags.isEmpty {
+                let section = tableData.createNewSection()
+                section.headerText = localizedString(header)
+                for tagDict in filteredTags {
+                    let row = section.createNewRow()
+                    row.cellType = OASimpleTableViewCell.getIdentifier()
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift:52]: reuse

File: Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                    row.descr = tagDict["localizedTitle"] as? String ?? ""
+                }
+            }
+        }
+    }
+    
+    override func getRow(_ indexPath: IndexPath?) -> UITableViewCell? {
+        guard let indexPath else { return nil }
+        let item = tableData.item(for: indexPath)
+        let cell = tableView.dequeueReusableCell(withIdentifier: OASimpleTableViewCell.getIdentifier(), for: indexPath) as! OASimpleTableViewCell
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/NameTagsDetailsViewController.swift:64]: reuseIdentifier

File: Sources/Controllers/TargetMenu/OARowInfo.h [hunk 8]
```diff
@@ -38,12 +38,14 @@
 
 @property (nonatomic) int height;
 @property (nonatomic) BOOL moreText;
+@property (nonatomic, strong) NSMutableArray<NSDictionary *> *detailsArray;
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/OARowInfo.h:41]: readonly

File: Sources/Controllers/TargetMenu/OARowInfo.h [hunk 9]
```diff
@@ -38,12 +38,14 @@
...
 
 @property (weak, nonatomic) id<OARowInfoDelegate> delegate;
 
 - (instancetype) initWithKey:(NSString *)key icon:(UIImage *)icon textPrefix:(NSString *)textPrefix text:(NSString *)text textColor:(UIColor *)textColor isText:(BOOL)isText needLinks:(BOOL)needLinks order:(int)order typeName:(NSString *)typeName isPhoneNumber:(BOOL)isPhoneNumber isUrl:(BOOL)isUrl;
 
 - (int) getRawHeight;
 - (UIFont *) getFont;
+- (void)setDetailsArray:(NSMutableArray<NSDictionary *> *)detailsArray;
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/OARowInfo.h:49]: nonnull

File: Sources/Controllers/TargetMenu/POI/OAPOIViewController.mm [hunk 10]
```diff
@@ -772,6 +807,33 @@ - (NSString *)formatPrefix:(NSString *)prefix units:(NSString *)units
     return prefix != nil && prefix.length > 0 ? [NSString stringWithFormat:@"%@, %@", prefix, units] : units;
 }
 
+- (NSArray<NSDictionary *> *)getPoiTypeDataForKey:(NSString *)routeTagKey withValue:(NSString *)value
```
Comment (tigrim) [inline Sources/Controllers/TargetMenu/POI/OAPOIViewController.mm:810]: duplicate?

**Review by tigrim — APPROVED** [review]

File: Sources/POI/OAPOIHelper.mm [hunk 11]
```diff
@@ -1503,6 +1472,17 @@ - (BOOL) breakSearch
     return _breakSearch;
 }
 
+- (BOOL) shouldProcessNameTagForKey:(NSString *)key
```
Comment (alex-osm) [inline Sources/POI/OAPOIHelper.mm:1475]: `- (BOOL) isNameTag:(NSString *)tag`

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 12]
```diff
@@ -229,6 +227,20 @@ - (void) populateRouteInfoSection:(OAGPXTableData *)data routeKey:(OARouteKey *)
         else if ([routeTagKey isEqualToString:@"wikipedia"])
             routeTagValue = [OAWikiAlgorithms getWikiUrlWithText:routeTagValue];
 
+        if ([[OAPOIHelper sharedInstance] shouldProcessNameTagForKey:routeTagKey] && !hasName)
```
Comment (alex-osm) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:230]: Check `!hasName` first

File: Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm [hunk 13]
```diff
@@ -223,6 +227,20 @@ - (void) populateRouteInfoSection:(OAGPXTableData *)data routeKey:(OARouteKey *)
...
+        {
+            OAGPXTableCellData *routeNameCellData = [OAGPXTableCellData withData:@{
+                kTableKey: @"name",
+                kCellType: [OAValueTableViewCell getCellIdentifier],
+                kCellTitle: OALocalizedString(@"shared_string_name"),
+                kCellDesc: (nameTagString.length > 0) ? nameTagString : i.value().toNSString(),
```
Comment (alex-osm) [inline Sources/Controllers/TargetMenu/GPX/TrackMenu/MenuBuilder/OATrackMenuTabOverview.mm:236]: is this logic correct?

**Pull request merged**